master/all/man/man3/KbBind.3
.TH "KbBind" 3 "Sat Jan 20 2018" "Version v0.2.8 at branch master" "ckb-next" \" -*- nroff -*-
.ad l
.nh
.SH NAME
KbBind \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <src/ckb/kbbind\&.h>\fP
.PP
Inherits \fBQObject\fP\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBkeyEvent\fP (const QString &\fBkey\fP, bool down)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBdidLoad\fP ()"
.br
.ti -1c
.RI "void \fBlayoutChanged\fP ()"
.br
.ti -1c
.RI "void \fBupdated\fP ()"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBKbBind\fP (\fBKbMode\fP *\fBmodeParent\fP, \fBKb\fP *\fBdevParent\fP, const \fBKeyMap\fP &keyMap)"
.br
.ti -1c
.RI "\fBKbBind\fP (\fBKbMode\fP *\fBmodeParent\fP, \fBKb\fP *\fBdevParent\fP, const \fBKeyMap\fP &keyMap, const \fBKbBind\fP &other)"
.br
.RI "\fI\fBKbBind\fP\&. \fP"
.ti -1c
.RI "void \fBload\fP (\fBCkbSettings\fP &settings)"
.br
.ti -1c
.RI "void \fBsave\fP (\fBCkbSettings\fP &settings)"
.br
.ti -1c
.RI "bool \fBneedsSave\fP () const "
.br
.ti -1c
.RI "const \fBKeyMap\fP & \fBmap\fP ()"
.br
.ti -1c
.RI "void \fBmap\fP (const \fBKeyMap\fP &map)"
.br
.ti -1c
.RI "bool \fBisISO\fP () const "
.br
.ti -1c
.RI "bool \fBisKeyboard\fP () const "
.br
.ti -1c
.RI "bool \fBisMouse\fP () const "
.br
.ti -1c
.RI "\fBKbPerf\fP * \fBperf\fP ()"
.br
.ti -1c
.RI "\fBKbLight\fP * \fBlight\fP ()"
.br
.ti -1c
.RI "QString \fBaction\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "QString \fBdefaultAction\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "QString \fBfriendlyName\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "QString \fBfriendlyActionName\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "void \fBsetAction\fP (const QString &\fBkey\fP, const QString &\fBaction\fP)"
.br
.ti -1c
.RI "void \fBsetAction\fP (const QStringList &keys, const QString &\fBaction\fP)"
.br
.ti -1c
.RI "void \fBresetAction\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "void \fBresetAction\fP (const QStringList &keys)"
.br
.ti -1c
.RI "void \fBnoAction\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "void \fBnoAction\fP (const QStringList &keys)"
.br
.ti -1c
.RI "bool \fBwinLock\fP ()"
.br
.ti -1c
.RI "void \fBwinLock\fP (bool newWinLock)"
.br
.ti -1c
.RI "void \fBupdate\fP (QFile &\fBcmd\fP, bool force=false)"
.br
.ti -1c
.RI "void \fBsetNeedsUpdate\fP ()"
.br
.ti -1c
.RI "int \fBgetMacroNumber\fP ()"
.br
.RI "\fI\fBKbBind::getMacroNumber\fP\&. \fP"
.ti -1c
.RI "QString \fBgetMacroPath\fP ()"
.br
.RI "\fI\fBKbBind::getMacroPath\fP\&. \fP"
.ti -1c
.RI "void \fBhandleNotificationChannel\fP (bool start)"
.br
.RI "\fIhandleNotificationChannel sends commands to ckb-daemon for (de-) activating the notify channel\&. Send a notify cmd to the keyboard to set or clear notification for reading macro definition\&. The file handle for the cmd pipe is stored in lastCmd\&. \fP"
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static QString \fBglobalRemap\fP (const QString &\fBkey\fP)"
.br
.ti -1c
.RI "static void \fBsetGlobalRemap\fP (const QHash< QString, QString > keyToActual)"
.br
.ti -1c
.RI "static void \fBloadGlobalRemap\fP ()"
.br
.ti -1c
.RI "static void \fBsaveGlobalRemap\fP ()"
.br
.ti -1c
.RI "static QString \fBdefaultAction\fP (const QString &\fBkey\fP, \fBKeyMap::Model\fP model)"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBKb\fP * \fBdevParent\fP () const "
.br
.ti -1c
.RI "\fBKbMode\fP * \fBmodeParent\fP () const "
.br
.ti -1c
.RI "\fBKeyAction\fP * \fBbindAction\fP (const QString &\fBkey\fP)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBKb\fP * \fB_devParent\fP"
.br
.ti -1c
.RI "quint64 \fBlastGlobalRemapTime\fP"
.br
.ti -1c
.RI "QFile * \fBlastCmd\fP"
.br
.RI "\fIlastCmd is a cache-hack\&. Because the QFile ist opened in \fBKb\fP, and we need it in the macro processing functions, we cache the value her in lastCmd\&. \fP"
.ti -1c
.RI "\fBKeyMap\fP \fB_map\fP"
.br
.ti -1c
.RI "QHash< QString, \fBKeyAction\fP * > \fB_bind\fP"
.br
.ti -1c
.RI "bool \fB_winLock\fP"
.br
.ti -1c
.RI "bool \fB_needsUpdate\fP"
.br
.ti -1c
.RI "bool \fB_needsSave\fP"
.br
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static QHash< QString, QString > \fB_globalRemap\fP"
.br
.ti -1c
.RI "static quint64 \fBglobalRemapTime\fP = 0"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "class \fBKeyAction\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 20 of file kbbind\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "KbBind::KbBind (\fBKbMode\fP *modeParent, \fBKb\fP *devParent, const \fBKeyMap\fP &keyMap)\fC [explicit]\fP"

.PP
Definition at line 12 of file kbbind\&.cpp\&.
.PP
.nf
12                                                                         :
13     QObject(modeParent), _devParent(parentBoard), lastGlobalRemapTime(globalRemapTime), _map(keyMap),
14     _winLock(false), _needsUpdate(true), _needsSave(true) {
15 }
.fi
.SS "KbBind::KbBind (\fBKbMode\fP *modeParent, \fBKb\fP *parentBoard, const \fBKeyMap\fP &keyMap, const \fBKbBind\fP &other)"
\fBKbBind::KbBind\fP // copy all existing \fBKey\fP bindings\&.
.PP
\fBParameters:\fP
.RS 4
\fImodeParent\fP 
.br
\fIdevParent\fP 
.br
\fIkeyMap\fP 
.br
\fIother\fP This is the \fBKbBind\fP object to copy from Use this constructor to copy an existing Binding
.br
\fImodeParent\fP 
.br
\fIparentBoard\fP 
.br
\fIkeyMap\fP 
.br
\fIother\fP 
.RE
.PP
Create a new Hash table and copy all entries
.PP
clear the destination list (there are the original KeyActions as references, so do not delete them)
.PP
and move the KeyActions we just created 
.PP
Definition at line 24 of file kbbind\&.cpp\&.
.PP
References _bind, KeyAction, map(), and KeyAction::value()\&.
.PP
.nf
24                                                                                              :
25     QObject(modeParent), _devParent(parentBoard), lastGlobalRemapTime(globalRemapTime), _bind(other\&._bind),
26     _winLock(false), _needsUpdate(true), _needsSave(true) {
27     map(keyMap);
28 
30     QHash<QString, KeyAction*> newBind;
31     foreach(QString key, _bind\&.keys()) {
32         KeyAction* act = _bind\&.value(key);
33         if(act) {
34             newBind[key] = new KeyAction(act->value(), this);
35         }
36     }
37 
39     _bind\&.clear();
40     foreach(QString key, newBind\&.keys()) {
41         KeyAction* act = newBind\&.value(key);
42         if(act) {
44             _bind[key] = new KeyAction(act->value(), this);
45         }
46     }
47     newBind\&.clear();      // here we *must not* delete the KeyActions, because they are referenced by _bind now
48 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "QString KbBind::action (const QString &key)"

.PP
Definition at line 145 of file kbbind\&.cpp\&.
.PP
References bindAction(), globalRemap(), and KeyAction::value()\&.
.PP
Referenced by friendlyActionName(), KbBindWidget::on_copyButton_clicked(), resetAction(), RebindWidget::setSelection(), and KbBindWidget::updateBind()\&.
.PP
.nf
145                                         {
146     QString rKey = globalRemap(key);
147     return bindAction(rKey)->value();
148 }
.fi
.SS "\fBKeyAction\fP * KbBind::bindAction (const QString &key)\fC [private]\fP"

.PP
Definition at line 140 of file kbbind\&.cpp\&.
.PP
References _bind, KeyAction::defaultAction(), devParent(), and KeyAction\&.
.PP
Referenced by action(), keyEvent(), and update()\&.
.PP
.nf
140                                                 {
141   if(!_bind\&.contains(key)) return _bind[key] = new KeyAction(KeyAction::defaultAction(key, devParent()->model()), this);
142   return _bind[key];
143 }
.fi
.SS "QString KbBind::defaultAction (const QString &key)"

.PP
Definition at line 155 of file kbbind\&.cpp\&.
.PP
References devParent()\&.
.PP
Referenced by KeyWidget::paintEvent(), and RebindWidget::setBind()\&.
.PP
.nf
155                                                {
156   return defaultAction(key, devParent()->model());
157 }
.fi
.SS "QString KbBind::defaultAction (const QString &key, \fBKeyMap::Model\fPmodel)\fC [static]\fP"

.PP
Definition at line 150 of file kbbind\&.cpp\&.
.PP
References KeyAction::defaultAction(), and globalRemap()\&.
.PP
.nf
150                                                                   {
151     QString rKey = globalRemap(key);
152     return KeyAction::defaultAction(rKey, model);
153 }
.fi
.SS "\fBKb\fP* KbBind::devParent () const\fC [inline]\fP, \fC [private]\fP"

.PP
Definition at line 115 of file kbbind\&.h\&.
.PP
References _devParent\&.
.PP
Referenced by bindAction(), defaultAction(), getMacroNumber(), getMacroPath(), KeyAction::keyEvent(), and save()\&.
.PP
.nf
115 { return _devParent; }
.fi
.SS "void KbBind::didLoad ()\fC [signal]\fP"

.PP
Definition at line 155 of file moc_kbbind\&.cpp\&.
.PP
Referenced by load()\&.
.PP
.nf
156 {
157     QMetaObject::activate(this, &staticMetaObject, 0, Q_NULLPTR);
158 }
.fi
.SS "QString KbBind::friendlyActionName (const QString &key)"

.PP
Definition at line 166 of file kbbind\&.cpp\&.
.PP
References _map, action(), and KeyAction\&.
.PP
Referenced by KbBindWidget::updateSelDisplay()\&.
.PP
.nf
166                                                     {
167     QString act = action(key);
168     return KeyAction(act)\&.friendlyName(_map);
169 }
.fi
.SS "QString KbBind::friendlyName (const QString &key)"

.PP
Definition at line 159 of file kbbind\&.cpp\&.
.PP
References _map, Key::friendlyName(), and globalRemap()\&.
.PP
.nf
159                                               {
160     const Key& pos = _map[globalRemap(key)];
161     if(!pos)
162         return "(Unknown)";
163     return pos\&.friendlyName();
164 }
.fi
.SS "int KbBind::getMacroNumber ()"

.PP
\fBReturns:\fP
.RS 4
number of notification channel\&. Use it in combination with notifyon/off-Statement 
.RE
.PP

.PP
Definition at line 266 of file kbbind\&.cpp\&.
.PP
References devParent(), and Kb::getMacroNumber()\&.
.PP
Referenced by handleNotificationChannel(), and RebindWidget::on_btnStartMacro_clicked()\&.
.PP
.nf
266                            {
267     return devParent()->getMacroNumber();
268 }
.fi
.SS "QString KbBind::getMacroPath ()"

.PP
\fBReturns:\fP
.RS 4
Filepath of macro notification pipe\&. If not set, returns initial value '' 
.RE
.PP

.PP
Definition at line 274 of file kbbind\&.cpp\&.
.PP
References devParent(), and Kb::getMacroPath()\&.
.PP
Referenced by RebindWidget::on_btnStartMacro_clicked()\&.
.PP
.nf
274                              {
275     return devParent()->getMacroPath();
276 }
.fi
.SS "QString KbBind::globalRemap (const QString &key)\fC [static]\fP"

.PP
Definition at line 98 of file kbbind\&.cpp\&.
.PP
References _globalRemap\&.
.PP
Referenced by action(), defaultAction(), friendlyName(), keyEvent(), noAction(), KeyWidget::paintEvent(), resetAction(), setAction(), and SettingsWidget::SettingsWidget()\&.
.PP
.nf
98                                              {
99     if(!_globalRemap\&.contains(key))
100         return key;
101     return _globalRemap\&.value(key);
102 }
.fi
.SS "void KbBind::handleNotificationChannel (boolstart)"

.PP
\fBParameters:\fP
.RS 4
\fIstart\fP is boolean\&. If true, notification channel is opened for all keys, otherwise channel ist closed\&.
.br
\fIstart\fP If true, notification channel is opened for all keys, otherwise channel ist closed\&. 
.RE
.PP

.PP
Definition at line 284 of file kbbind\&.cpp\&.
.PP
References getMacroNumber(), and lastCmd\&.
.PP
Referenced by RebindWidget::on_btnStartMacro_clicked(), and RebindWidget::on_btnStopMacro_clicked()\&.
.PP
.nf
284                                                  {
285     if (getMacroNumber() > 0 && lastCmd) {
286         if (start) {
287             lastCmd->write (QString("\nnotifyon %1\n@%1 notify all:on\n")\&.arg(getMacroNumber())\&.toLatin1());
288         } else {
289             lastCmd->write (QString("\n@%1 notify all:off\nnotifyoff %1\n")\&.arg(getMacroNumber())\&.toLatin1());
290         }
291         lastCmd->flush();
292     } else qDebug() << QString("No cmd or valid handle for notification found, macroNumber = %1, lastCmd = %2")
293                        \&.arg(getMacroNumber())\&.arg(lastCmd? "set" : "unset");
294 }
.fi
.SS "bool KbBind::isISO () const\fC [inline]\fP"

.PP
Definition at line 43 of file kbbind\&.h\&.
.PP
References _map, and KeyMap::isISO()\&.
.PP
.nf
43 { return _map\&.isISO(); }
.fi
.SS "bool KbBind::isKeyboard () const\fC [inline]\fP"

.PP
Definition at line 44 of file kbbind\&.h\&.
.PP
References _map, and KeyMap::isKeyboard()\&.
.PP
.nf
44 { return _map\&.isKeyboard(); }
.fi
.SS "bool KbBind::isMouse () const\fC [inline]\fP"

.PP
Definition at line 45 of file kbbind\&.h\&.
.PP
References _map, and KeyMap::isMouse()\&.
.PP
Referenced by KbBindWidget::on_copyButton_clicked(), KbBindWidget::on_resetButton_clicked(), RebindWidget::setBind(), and KbBindWidget::updateSelDisplay()\&.
.PP
.nf
45 { return _map\&.isMouse(); }
.fi
.SS "void KbBind::keyEvent (const QString &key, booldown)\fC [slot]\fP"

.PP
Definition at line 296 of file kbbind\&.cpp\&.
.PP
References bindAction(), globalRemap(), and KeyAction::keyEvent()\&.
.PP
Referenced by Kb::readNotify()\&.
.PP
.nf
296                                                   {
297     QString rKey = globalRemap(key);
298     KeyAction* act = bindAction(rKey);
299     if(act)
300         act->keyEvent(this, down);
301 }
.fi
.SS "void KbBind::layoutChanged ()\fC [signal]\fP"

.PP
Definition at line 161 of file moc_kbbind\&.cpp\&.
.PP
Referenced by map()\&.
.PP
.nf
162 {
163     QMetaObject::activate(this, &staticMetaObject, 1, Q_NULLPTR);
164 }
.fi
.SS "\fBKbLight\fP * KbBind::light ()"

.PP
Definition at line 54 of file kbbind\&.cpp\&.
.PP
References KbMode::light(), and modeParent()\&.
.PP
Referenced by RebindWidget::applyChanges(), KeyAction::keyEvent(), and RebindWidget::setSelection()\&.
.PP
.nf
54                       {
55     return modeParent()->light();
56 }
.fi
.SS "void KbBind::load (\fBCkbSettings\fP &settings)"

.PP
Definition at line 58 of file kbbind\&.cpp\&.
.PP
References _bind, _map, _needsSave, _winLock, CkbSettings::childKeys(), didLoad(), KeyMap::fromName(), KeyMap::fromStorage(), KeyAction, map(), and CkbSettings::value()\&.
.PP
Referenced by KbMode::KbMode()\&.
.PP
.nf
58                                       {
59     _needsSave = false;
60     SGroup group(settings, "Binding");
61     KeyMap currentMap = _map;
62     _map = KeyMap::fromName(settings\&.value("KeyMap")\&.toString());
63     // Load key settings
64     bool useReal = settings\&.value("UseRealNames")\&.toBool();
65     _bind\&.clear();
66     {
67         SGroup group(settings, "Keys");
68         foreach(QString key, settings\&.childKeys()){
69             QString name = key\&.toLower();
70             if(!useReal)
71                 name = _map\&.fromStorage(name);
72             QString bind = settings\&.value(key)\&.toString();
73             _bind[name] = new KeyAction(bind, this);
74         }
75     }
76     _winLock = settings\&.value("WinLock")\&.toBool();
77     emit didLoad();
78     map(currentMap);
79 }
.fi
.SS "void KbBind::loadGlobalRemap ()\fC [static]\fP"

.PP
Definition at line 116 of file kbbind\&.cpp\&.
.PP
References _globalRemap, CkbSettings::childKeys(), globalRemapTime, and CkbSettings::value()\&.
.PP
Referenced by SettingsWidget::SettingsWidget()\&.
.PP
.nf
116                             {
117     _globalRemap\&.clear();
118     CkbSettings settings("Program/GlobalRemap");
119     foreach(const QString& key, settings\&.childKeys())
120         _globalRemap[key] = settings\&.value(key)\&.toString();
121     globalRemapTime = QDateTime::currentMSecsSinceEpoch();
122 }
.fi
.SS "const \fBKeyMap\fP& KbBind::map ()\fC [inline]\fP"

.PP
Definition at line 41 of file kbbind\&.h\&.
.PP
References _map\&.
.PP
Referenced by KbBind(), KbMode::keyMap(), load(), map(), KbBindWidget::newLayout(), KbBindWidget::on_copyButton_clicked(), KbBindWidget::on_resetButton_clicked(), RebindWidget::setBind(), KbBindWidget::updateBind(), and KbBindWidget::updateSelDisplay()\&.
.PP
.nf
41 { return _map; }
.fi
.SS "void KbBind::map (const \fBKeyMap\fP &map)"

.PP
Definition at line 133 of file kbbind\&.cpp\&.
.PP
References _map, _needsSave, _needsUpdate, layoutChanged(), and map()\&.
.PP
.nf
133                                  {
134     _map = map;
135     _needsUpdate = true;
136     _needsSave = true;
137     emit layoutChanged();
138 }
.fi
.SS "\fBKbMode\fP* KbBind::modeParent () const\fC [inline]\fP, \fC [private]\fP"

.PP
Definition at line 116 of file kbbind\&.h\&.
.PP
Referenced by light(), and perf()\&.
.PP
.nf
116 { return (KbMode*)parent(); }
.fi
.SS "bool KbBind::needsSave () const\fC [inline]\fP"

.PP
Definition at line 38 of file kbbind\&.h\&.
.PP
References _needsSave\&.
.PP
Referenced by KbMode::needsSave()\&.
.PP
.nf
38 { return _needsSave; }
.fi
.SS "void KbBind::noAction (const QString &key)"

.PP
Definition at line 181 of file kbbind\&.cpp\&.
.PP
References _bind, _map, globalRemap(), KeyMap::key(), KeyAction, and resetAction()\&.
.PP
Referenced by RebindWidget::applyChanges(), and RebindWidget::on_unbindButton_clicked()\&.
.PP
.nf
181                                        {
182     resetAction(key);
183     QString rKey = globalRemap(key);
184     if(!_map\&.key(rKey))
185         return;
186     _bind[rKey] = new KeyAction(this);
187 }
.fi
.SS "void KbBind::noAction (const QStringList &keys)\fC [inline]\fP"

.PP
Definition at line 76 of file kbbind\&.h\&.
.PP
References noAction()\&.
.PP
Referenced by noAction()\&.
.PP
.nf
76 { foreach(const QString& key, keys) noAction(key); }
.fi
.SS "\fBKbPerf\fP * KbBind::perf ()"

.PP
Definition at line 50 of file kbbind\&.cpp\&.
.PP
References modeParent(), and KbMode::perf()\&.
.PP
Referenced by KeyAction::keyEvent(), and RebindWidget::setBind()\&.
.PP
.nf
50                     {
51     return modeParent()->perf();
52 }
.fi
.SS "void KbBind::resetAction (const QString &key)"

.PP
Definition at line 171 of file kbbind\&.cpp\&.
.PP
References _bind, _needsSave, _needsUpdate, action(), and globalRemap()\&.
.PP
Referenced by noAction(), KbBindWidget::on_resetButton_clicked(), RebindWidget::on_resetButton_clicked(), and setAction()\&.
.PP
.nf
171                                           {
172     QString rKey = globalRemap(key);
173     // Clean up existing action (if any)
174     KeyAction* action = _bind\&.value(rKey);
175     delete action;
176     _bind\&.remove(rKey);
177     _needsUpdate = true;
178     _needsSave = true;
179 }
.fi
.SS "void KbBind::resetAction (const QStringList &keys)\fC [inline]\fP"

.PP
Definition at line 73 of file kbbind\&.h\&.
.PP
References resetAction()\&.
.PP
Referenced by resetAction()\&.
.PP
.nf
73 { foreach(const QString& key, keys) resetAction(key); }
.fi
.SS "void KbBind::save (\fBCkbSettings\fP &settings)"

.PP
Definition at line 81 of file kbbind\&.cpp\&.
.PP
References _bind, _map, _needsSave, _winLock, KeyAction::defaultAction(), devParent(), KeyMap::name(), CkbSettings::setValue(), and KeyAction::value()\&.
.PP
Referenced by KbMode::save()\&.
.PP
.nf
81                                       {
82     _needsSave = false;
83     SGroup group(settings, "Binding");
84     settings\&.setValue("KeyMap", _map\&.name());
85     // Save key settings
86     settings\&.setValue("UseRealNames", true);
87     {
88         SGroup group(settings, "Keys");
89         foreach(QString key, _bind\&.keys()){
90             KeyAction* act = _bind\&.value(key);
91             if(act && act->value() != KeyAction::defaultAction(key, devParent()->model()))
92                 settings\&.setValue(key, act->value());
93         }
94     }
95     settings\&.setValue("WinLock", _winLock);
96 }
.fi
.SS "void KbBind::saveGlobalRemap ()\fC [static]\fP"

.PP
Definition at line 124 of file kbbind\&.cpp\&.
.PP
References _globalRemap, and CkbSettings::setValue()\&.
.PP
Referenced by SettingsWidget::updateModifiers()\&.
.PP
.nf
124                             {
125     CkbSettings settings("Program/GlobalRemap", true);
126     QHashIterator<QString, QString> i(_globalRemap);
127     while(i\&.hasNext()){
128         i\&.next();
129         settings\&.setValue(i\&.key(), i\&.value());
130     }
131 }
.fi
.SS "void KbBind::setAction (const QString &key, const QString &action)"

.PP
Definition at line 189 of file kbbind\&.cpp\&.
.PP
References _bind, _map, globalRemap(), KeyMap::key(), KeyAction, and resetAction()\&.
.PP
Referenced by RebindWidget::applyChanges(), and KbBindWidget::on_copyButton_clicked()\&.
.PP
.nf
189                                                                {
190     resetAction(key);

src/ckb/kbbind.cpp
#include <fcntl.h>
#include <QDateTime>
#include <QUrl>
#include "ckbsettings.h"
#include "kbbind.h"
#include "kbmode.h"
#include "kb.h"
#include "qdebug.h"

QHash<QString, QString> KbBind::_globalRemap;
quint64 KbBind::globalRemapTime = 0;

KbBind::KbBind(KbMode* modeParent, Kb* parentBoard, const KeyMap& keyMap) :
    QObject(modeParent), _devParent(parentBoard), lastGlobalRemapTime(globalRemapTime), _map(keyMap),
    _winLock(false), _needsUpdate(true), _needsSave(true) {
}

///
/// \brief KbBind::KbBind   // copy all existing Key bindings
/// \param modeParent
/// \param parentBoard
/// \param keyMap
/// \param other
///
KbBind::KbBind(KbMode* modeParent, Kb* parentBoard, const KeyMap& keyMap, const KbBind& other) :
    QObject(modeParent), _devParent(parentBoard), lastGlobalRemapTime(globalRemapTime), _bind(other._bind),
    _winLock(false), _needsUpdate(true), _needsSave(true) {
    map(keyMap);

    /// Create a new Hash table and copy all entries
    QHash<QString, KeyAction*> newBind;
    foreach(QString key, _bind.keys()) {
        KeyAction* act = _bind.value(key);
        if(act) {
            newBind[key] = new KeyAction(act->value(), this);
        }
    }

    /// clear the destination list (there are the original KeyActions as references, so do not delete them)
    _bind.clear();
    foreach(QString key, newBind.keys()) {
        KeyAction* act = newBind.value(key);
        if(act) {
            /// and move the KeyActions we just created
            _bind[key] = new KeyAction(act->value(), this);
        }
    }
    newBind.clear();      // here we *must not* delete the KeyActions, because they are referenced by _bind now
}

KbPerf* KbBind::perf(){
    return modeParent()->perf();
}

KbLight* KbBind::light(){
    return modeParent()->light();
}

void KbBind::load(CkbSettings& settings){
    _needsSave = false;
    SGroup group(settings, "Binding");
    KeyMap currentMap = _map;
    _map = KeyMap::fromName(settings.value("KeyMap").toString());
    // Load key settings
    bool useReal = settings.value("UseRealNames").toBool();
    _bind.clear();
    {
        SGroup group(settings, "Keys");
        foreach(QString key, settings.childKeys()){
            QString name = key.toLower();
            if(!useReal)
                name = _map.fromStorage(name);
            QString bind = settings.value(key).toString();
            _bind[name] = new KeyAction(bind, this);
        }
    }
    _winLock = settings.value("WinLock").toBool();
    emit didLoad();
    map(currentMap);
}

void KbBind::save(CkbSettings& settings){
    _needsSave = false;
    SGroup group(settings, "Binding");
    settings.setValue("KeyMap", _map.name());
    // Save key settings
    settings.setValue("UseRealNames", true);
    {
        SGroup group(settings, "Keys");
        foreach(QString key, _bind.keys()){
            KeyAction* act = _bind.value(key);
            if(act && act->value() != KeyAction::defaultAction(key))
                settings.setValue(key, act->value());
        }
    }
    settings.setValue("WinLock", _winLock);
}

QString KbBind::globalRemap(const QString& key){
    if(!_globalRemap.contains(key))
        return key;
    return _globalRemap.value(key);
}

void KbBind::setGlobalRemap(const QHash<QString, QString> keyToActual){
    _globalRemap.clear();
    // Ignore any keys with the standard binding
    QHashIterator<QString, QString> i(keyToActual);
    while(i.hasNext()){
        i.next();
        if(i.key() != i.value())
            _globalRemap[i.key()] = i.value();
    }
    globalRemapTime = QDateTime::currentMSecsSinceEpoch();
}

void KbBind::loadGlobalRemap(){
    _globalRemap.clear();
    CkbSettings settings("Program/GlobalRemap");
    foreach(const QString& key, settings.childKeys())
        _globalRemap[key] = settings.value(key).toString();
    globalRemapTime = QDateTime::currentMSecsSinceEpoch();
}

void KbBind::saveGlobalRemap(){
    CkbSettings settings("Program/GlobalRemap", true);
    QHashIterator<QString, QString> i(_globalRemap);
    while(i.hasNext()){
        i.next();
        settings.setValue(i.key(), i.value());
    }
}

void KbBind::map(const KeyMap& map){
    _map = map;
    _needsUpdate = true;
    _needsSave = true;
    emit layoutChanged();
}

QString KbBind::action(const QString& key){
    QString rKey = globalRemap(key);
    return bindAction(rKey)->value();
}

QString KbBind::defaultAction(const QString& key){
    QString rKey = globalRemap(key);
    return KeyAction::defaultAction(rKey);
}

QString KbBind::friendlyName(const QString& key){
    const Key& pos = _map[globalRemap(key)];
    if(!pos)
        return "(Unknown)";
    return pos.friendlyName();
}

QString KbBind::friendlyActionName(const QString& key){
    QString act = action(key);
    return KeyAction(act).friendlyName(_map);
}

void KbBind::resetAction(const QString &key){
    QString rKey = globalRemap(key);
    // Clean up existing action (if any)
    KeyAction* action = _bind.value(rKey);
    delete action;
    _bind.remove(rKey);
    _needsUpdate = true;
    _needsSave = true;
}

void KbBind::noAction(const QString& key){
    resetAction(key);
    QString rKey = globalRemap(key);
    if(!_map.key(rKey))
        return;
    _bind[rKey] = new KeyAction(this);
}

void KbBind::setAction(const QString& key, const QString& action){
    resetAction(key);
    QString rKey = globalRemap(key);
    if(!_map.key(rKey))
        return;
    _bind[rKey] = new KeyAction(action, this);
}

void KbBind::update(QFile& cmd, bool force){
    if(!force && !_needsUpdate && lastGlobalRemapTime == globalRemapTime)
        return;
    lastGlobalRemapTime = globalRemapTime;
    emit updated();
    _needsUpdate = false;
    // Reset all keys and enable notifications for all
    cmd.write("rebind all notify all");
    // Make sure modifier keys are included as they may be remapped globally
    QHash<QString, KeyAction*> bind(_bind);
    if(!_bind.contains("caps")) bind["caps"] = 0;
    if(!_bind.contains("lshift")) bind["lshift"] = 0;
    if(!_bind.contains("rshift")) bind["rshift"] = 0;
    if(!_bind.contains("lctrl")) bind["lctrl"] = 0;
    if(!_bind.contains("rctrl")) bind["rctrl"] = 0;
    if(!_bind.contains("lwin")) bind["lwin"] = 0;
    if(!_bind.contains("rwin")) bind["rwin"] = 0;
    if(!_bind.contains("lalt")) bind["lalt"] = 0;
    if(!_bind.contains("ralt")) bind["ralt"] = 0;
    if(!_bind.contains("fn")) bind["fn"] = 0;
    QHashIterator<QString, KeyAction*> i(bind);

    // Initialize String buffer for macro Key definitions (G-keys)
    // "macro clear" is neccessary, if an older definition is unbound.
    QString macros = "\nmacro clear\n";

    // Write out rebound keys and collect infos for macro definitions
    while(i.hasNext()){
        i.next();
        QString key = i.key();
        KeyAction* act = i.value();
        if(_globalRemap.contains(key))
            act = bindAction(_globalRemap.value(key));
        if(!act)
            continue;
        QString value = act->driverName();
        if(value.isEmpty()){
            // If the key is unbound or is a special action, unbind it
            cmd.write(" unbind ");
            cmd.write(key.toLatin1());
            // if a macro definiton for the key is given,
            // add the converted string to key-buffer "macro"
            if (act->isValidMacro()) {
                if (act->macroContent().length() > 0) {
                    macros.append("macro " + key.toLatin1() + ":" + act->macroContent().toLatin1() + "\n");
                }
            }
        } else {
            // Otherwise, write the binding
            cmd.write(" bind ");
            cmd.write(key.toLatin1());
            cmd.write(":");
            cmd.write(value.toLatin1());
        }
    }
    // If win lock is enabled, unbind windows keys
    if(_winLock)
        cmd.write(" unbind lwin rwin");

    // At last, send Macro definitions if available.
    // If no definitions are made, clear macro will be sent only to reset all macros,
    cmd.write(macros.toLatin1());
    lastCmd = &cmd;
}

////////
/// \brief KbBind::getMacroNumber
/// \return number of notification channel. Use it in combination with notifyon/off-Statement
///
int KbBind::getMacroNumber() {
    return devParent()->getMacroNumber();
}

////////
/// \brief KbBind::getMacroPath
/// \return Filepath of macro notification pipe. If not set, returns initial value ""
///
QString KbBind::getMacroPath() {
    return devParent()->getMacroPath();
}

////////
/// \brief handleNotificationChannel sends commands to ckb-daemon for (de-) activating the notify channel.
/// \param start If true, notification channel is opened for all keys, otherwise channel ist closed.
/// \param cmdpath QString sets the first part of the cmd pipe pathname (eg "/dev/input/ckb2")
/// to locate the correct keyboard related cmd pipe.
///
void KbBind::handleNotificationChannel(bool start, QString cmdpath) {
    if (getMacroNumber() > 0) {
        QFile cmd;
        cmdpath.append("/cmd");
        int fd = open(cmdpath.toLatin1().constData(), O_WRONLY | O_NONBLOCK);
        if(cmd.open(fd, QIODevice::WriteOnly, QFileDevice::AutoCloseHandle)) {
            qDebug() << "Writing to new Notification channel" << cmdpath << "with" << start;
            if (start) {
                cmd.write (QString("\nnotifyon %1\n@%1 notify all:on\n").arg(getMacroNumber()).toLatin1());
            } else {
                cmd.write(QString("\n@%1 notify all:off\nnotifyoff %1\n").arg(getMacroNumber()).toLatin1());
            }
            cmd.flush();
        } else {
            qDebug() << "error while opening" << cmdpath;
        }
        cmd.close();
    } else qDebug() << QString("No cmd or valid handle for notification found, macroNumber = %1, lastCmd = %2")
                       .arg(getMacroNumber()).arg(cmdpath);
}

///
/// \brief KbBind::keyEvent
/// \param key
/// \param down
///
void KbBind::keyEvent(const QString& key, bool down){
    QString rKey = globalRemap(key);
    KeyAction* act = bindAction(rKey);
    if(act)
        act->keyEvent(this, down);
}
